crates/autopilot/src/domain/settlement/mod.rs

```diff
@@ -28,6 +28,7 @@ pub struct Settlement {
     solution: Solution,
     transaction: Transaction,
     auction: Auction,
+    competition: domain::Competition,
```

Comment:
Why is competition a single solver, score, block? Isn't `competition` the set of all solutions that were submitted? Would those fields make more sense on `Solution`? Or can a solution exist without that information?

Reply:
> Why is competition a single solver, score, block?

Winner and winner score are outputs of the competition. Block is also more related to competition that it is related to `settlement::Auction`, and we also keep `settlement::Auction` more 1 to 1 with `domain::Auction` so we can merge them eventually.

>  Isn't competition the set of all solutions that were submitted

True. However we don't need this info to support current functionalities of OnSettlementEventUpdater. Once we start implementing circuit breaker, all competition related data is supposed to be put under `domain::Competition` (solutions (if we need them), winner/s etc).

> Would those fields make more sense on Solution? Or can a solution exist without that information?

If you mean `settlement::Solution`, no. `settlement::Solution` is an entity on it's own that shouldn't be related to anything in the competition, otherwise it would be completely non-reusable. `settlement::Auction` and `domain::Competition` are separated at the moment, as I am still not sure what their relation should be. Auction is input and Competition is output of the runloop iteration. Or we can think of Competition in a broader context in which it contains an Auction.

Reply:
> Winner and winner score are outputs of the competition.

They are also part of the solution though. `score()` is a function on `Solution` and `winner` is implicitly part of the transaction that the solution is built from. Why do we define a data object that has different ways of accessing the same thing?

I agree that deadline should probably not be on solution (and also not on auction, although it could be at least when we move to one block per batch). Until then, should we not have it as an isolated field on Settlement?

I really find the concept of "competition" for the three values that we defined here inappropriate.

Reply:
> They are also part of the solution though. score() is a function on Solution and winner is implicitly part of the transaction that the solution is built from. Why do we define a data object that has different ways of accessing the same thing?

I am not sure I understand this. `Settlement::Solution` has a `score()` function but that value has nothing to do with `domain::Competition::score` which is a driver reported score. Similarly, `Transaction::from` has nothing to do with `domain::Competition::winner` as those can be different as well (if non-winning solver settled a Settlement for example).

Maybe the name is not appropriate but the `domain::Competition` object represents/contains important data that is related to what happens at COMPETITION time. I am not happy about saving the `domain::Competition` as part of the `Settlement` so that might be confusing. But then, why we allowed a `settlement::Auction` to be part of the `Settlement`. Both `settlement::Auction` and `domain::Competition` are objects built and used at competition time and are associated to this `Settlement`.

Reply:
Unless you start with an assumption that all errors (like my example of non-winning solver settling a Settlement) are already handled in `Settlement` constructor so there is no need to save `winner` inside `Settlement` because that would be an error during construction of `Settlement`?

If so, then I need to implement this in this PR and completely skip saving `domain::Competition` inside `Settlement`.

Reply:
<img width="956" alt="Screenshot 2024-07-30 at 00 08 35" src="https://github.com/user-attachments/assets/d0b9ce66-1dd3-4fa6-bbeb-259da78049d4">

Is this approach you would prefer?

Reply:
> However, shouldn't the same then be true when auction is non-existent (ie. when a solver submitted a tx without referencing a valid auction)?

Current behaviour is to skip creating observation in this case: https://github.com/cowprotocol/services/blob/main/crates/autopilot/src/on_settlement_event_updater.rs#L156-L160

observation() is a function on Settlement just because it requires Transaction and Auction to be built. 

> which assumes valid auction

This wasn't my assumption from beginning, just because, as you mentioned, we sometimes need invalid settlements to work with. Instead, i planned to have a `Settlement` containing:
1. settlement calldata transformed to Solution because this is 99.9999% valid
2. fetch auction data
3. fetch competition data (solutions, winner, whatever happened at competition time).
4. fetch something else? like existence of settled orders in the database to detect if they are JIT orders etc.

Then, if we see that we have a need for `ValidSettlement` we can easily transform `Settlement` -> `ValidSettlement` that will have those guarantees you talked about.

Reply:
But why do we want to create inconsistent behaviour? If a solver misbehaves and submits an invalid settlement for whatever reason (lying about the auction id, lying about the score, lying about executions) should there be a settlement observation or not?

It sounds like without the auction we cannot create an "observation" simply because we cannot compute the score and fees without an auction. In this case, I think there should not be an observation in case a solver is lying about anything and thus `Settlement::new` can perform all required correctness checks. If in other places we want information about invalid transactions as well, we can still create a `Solution` from the transaction itself.

Reply:
Refactored, not sure if I addressed the original comment.